go/parquet/internal/encoding/encoding_test.go

```diff
@@ -0,0 +1,684 @@
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			enc := encoding.NewEncoder(parquet.Types.Int64, parquet.Encodings.DeltaBinaryPacked, false, column, memory.DefaultAllocator)
+
+			enc.(encoding.Int64Encoder).Put(tt.toencode)
+			buf, _ := enc.FlushValues()
+			defer buf.Release()
+
+			assert.Equal(t, tt.expected, buf.Bytes())
+
+			dec := encoding.NewDecoder(parquet.Types.Int64, parquet.Encodings.DeltaBinaryPacked, column, memory.DefaultAllocator)
+
+			dec.(encoding.Int64Decoder).SetData(len(tt.toencode), tt.expected)
+			out := make([]int64, len(tt.toencode))
+			dec.(encoding.Int64Decoder).Decode(out)
+			assert.Equal(t, tt.toencode, out)
+		})
+	}
+
+	t.Run("test progressive decoding", func(t *testing.T) {
+		values := make([]int64, 1000)
+		testutils.FillRandomInt64(0, values)
+
+		enc := encoding.NewEncoder(parquet.Types.Int64, parquet.Encodings.DeltaBinaryPacked, false, column, memory.DefaultAllocator)
+		enc.(encoding.Int64Encoder).Put(values)
+		buf, _ := enc.FlushValues()
+		defer buf.Release()
+
+		dec := encoding.NewDecoder(parquet.Types.Int64, parquet.Encodings.DeltaBinaryPacked, column, memory.DefaultAllocator)
+		dec.(encoding.Int64Decoder).SetData(len(values), buf.Bytes())
+
+		valueBuf := make([]int64, 100)
+		for i, j := 0, len(valueBuf); j <= len(values); i, j = i+len(valueBuf), j+len(valueBuf) {
+			decoded, _ := dec.(encoding.Int64Decoder).Decode(valueBuf)
+			assert.Equal(t, len(valueBuf), decoded)
+			assert.Equalf(t, values[i:j], valueBuf, "indexes %d:%d", i, j)
+		}
+	})
+}
+
+func TestDeltaLengthByteArrayEncoding(t *testing.T) {
+	column := schema.NewColumn(schema.NewByteArrayNode("bytearray", parquet.Repetitions.Required, -1), 0, 0)
+
+	test := []parquet.ByteArray{[]byte("Hello"), []byte("World"), []byte("Foobar"), []byte("ABCDEF")}
+	expected := []byte{128, 1, 4, 4, 10, 0, 1, 0, 0, 0, 2, 0, 0, 0, 72, 101, 108, 108, 111, 87, 111, 114, 108, 100, 70, 111, 111, 98, 97, 114, 65, 66, 67, 68, 69, 70}
+
+	enc := encoding.NewEncoder(parquet.Types.ByteArray, parquet.Encodings.DeltaLengthByteArray, false, column, memory.DefaultAllocator)
+	enc.(encoding.ByteArrayEncoder).Put(test)
+	buf, _ := enc.FlushValues()
+	defer buf.Release()
+
+	assert.Equal(t, expected, buf.Bytes())
+
+	dec := encoding.NewDecoder(parquet.Types.ByteArray, parquet.Encodings.DeltaLengthByteArray, column, nil)
+	dec.SetData(len(test), expected)
+	out := make([]parquet.ByteArray, len(test))
+	decoded, _ := dec.(encoding.ByteArrayDecoder).Decode(out)
+	assert.Equal(t, len(test), decoded)
+	assert.Equal(t, test, out)
+}
+
+func TestDeltaByteArrayEncoding(t *testing.T) {
+	test := []parquet.ByteArray{[]byte("Hello"), []byte("World"), []byte("Foobar"), []byte("ABCDEF")}
+	expected := []byte{128, 1, 4, 4, 0, 0, 0, 0, 0, 0, 128, 1, 4, 4, 10, 0, 1, 0, 0, 0, 2, 0, 0, 0, 72, 101, 108, 108, 111, 87, 111, 114, 108, 100, 70, 111, 111, 98, 97, 114, 65, 66, 67, 68, 69, 70}
```

Comment:
where do the expected values come from?

Reply:
The Delta Byte Array Encoding was implemented in another parquet library, so I stole the expected values from their unit tests and also did some manual confirmation to ensure they were the correct values to the best of my knowledge.